content/posts/a_confusing_build_failure.md

```diff
@@ -0,0 +1,64 @@
+could fail.
+
+These kind of tests give us no feedback and no confidence, in other words they provide zero value. We should either test the real thing or test it manually.
+
+## Summing up
+
+It's quite surprising how a single build failure unearthed multiple cultural and technical shortcomings. 
+
+ * We often push code that we broke - not only tests aren't green but also many times are not compiling. This means that tests are not treated as first class citizens.
+* Mocks are used ***too much***. Mocks are a special case of [test doubles](https://martinfowler.com/articles/mocksArentStubs.html) that are used for verification testing. They are a perfect tool for teasing out collaborators and interfaces, especially when the real objects are hard/costly to use, but should be used with moderation. This means a lack of understanding of the tools used for testing as well when they should be applied. 
+* No review on the test code is performed. This again means that tests are treated as second class citizens and a nice to have.
+* Local builds differ than CI builds. This means we're more prone to false positives/negatives.
+* Broken builds pass on the main trunk. This means that we're in a hurry more times than we would like.
+
+Broken builds should alert the whole team that something has gone bad and needs immediate attention and fix. Just having tests is not enough - they should provide fast feedback and confidence. They should be easy to read and understand, easy to create and their failures easy to decode, as well as deterministic. These qualities of tests are the weapons that a developer can use in order to refactor aggressively and keep the system easy to understand and change. Add
+critical thinking to the mix in order to listen to what tests are trying to tell us about our design and we have the ingredients to battle complexity. 
+
+This may sound a little provocative, but we could as well delete tests that do not adhere to the qualities above as we end up with double the burden of maintenance and confusion.
+
+As a personal critique on my part I could probably say that even though I tried, I should emphasize even more the dangers of mocking too much. After all one cannot learn just by one example, especially from code that is not created with testing in mind - meaning there will be cut corners and "broken" rules on these examples.
+
+Finally, this can be seen as a further proof that tools by themselves do not solve problems. Having tests and a CI server is not enough, it is needed to have a deep understanding why these set of techniques and tools are used, as well as the right place and way to use them. Fortunately these are not arcane magic but knowledge and insight that can be learned through repetitive practice and retrospective. We are bound to fail through inexperience but bound to succeed through
```

Comment:
```suggestion
Finally, this can be seen as a further proof that tools by themselves do not solve problems. Having tests and a CI server is not enough. It is critical to have a deep understanding why these set of techniques and tools are used, as well as the right place and way to use them. Fortunately, this level of understanding is not the result of arcane magic but can be acquired through repetitive practice and retrospection. We are bound to fail through inexperience but bound to succeed through
```